### PR TITLE
 [CU-r779ej] Add-token-validation

### PR DESCRIPTION
**Pull request by eastemployeer**
Validation is perfectly working. The way I'm signing in and validating is called "Authorization code flow" BUT with openid scope which provides me with necessary token for validation called id_token. Microsoft generates two types of token. First type is meant to be used only with MS Graph and not for validating our own API, that's why it is just not validatable (that's why we couldn't validate it earlier, we had just this type of token). Second type is token just for validating our own API and for nothing else. This one is called id_token. To get this and not ruin whole signing in implemented methodology I've added "openid" scope to acquire id_token alongside access_token (meant only for MS Graph). It is **possible** that some functionalites may not be working but it is easy fixable. I just don't remember if I re-added all admin consents that were added before (I had to delete some of them while testing).
Ref
 -> https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent
 -> https://docs.microsoft.com/pl-pl/azure/active-directory/develop/v2-oauth2-auth-code-flow
Zrzut ekranu z 2021-05-15 17-37-34
Especially last line is super important.

**Tl;dr;** Acquiring code is working in the same way that was working before. I've just added extra scope called 'openid' and that is generating extra token called id_token after `/token` endpoint is called (function `acquireTokenByCode` from MSAL). Then I am using this token for validation and the old token (We called it "jwt") for taking informations from Microsoft Graph.

**OTHER POSSIBLE SCENARIOS**
The way I'm doing it now is in my opinion the best way BUT I would like to describe other options, just for reference.
1. Auth code flow + openid scope --> while acquiring token we get extra token called id_token (solution that is implemented). Required small changes in MS Organization like allowing ID_tokens in Implicit Grant.
2. Auth code flow with  **ONLY** custom scope for our API (in our case it is 'api://btapp.com/besttournamentapp.read'). That is exactly this way: https://xsreality.medium.com/making-azure-ad-oidc-compliant-5734b70c43ff. But this solution has many disadvantages like: we are acquiring only token for validation our API. What with taking data from MS Organizations/Graph? That is the problem. To do this we have to acquire token **on behalf of the user**. That is bad solution because every time user wants to get some data from MS, backend has to acquire access token to MS Graph. Thats very unefficient.
3. Hybrid flow. This solution was interesting. It is described here: https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow. It gives us ID_token before we acquire access_token however tehere are other disadvantages (responseMode must be form_post, which cuts request-response handshake on backend and that is the problem because we have to then manually redirect to frontend again. That is looking bad to me. Other problem is that msal-node's functionalities are beyond this. I would have to implement it manually.

### CODE REVIEW

**Comment by Zummek** [comment]
Task linked: [CU-r779ej Add token validation](https://app.clickup.com/t/r779ej)

File: server/middlewares/authenticationMiddleware.ts [hunk 1]
```diff
@@ -1,63 +1,24 @@
...
-  const decoded = jwt.decode(token) as any;
-
-  if (!decoded) {
+  const token = req.cookies.validation_token;
+  const decoded: any = jwt.decode(token, { complete: true });
+  const key = await client.getSigningKey(decoded?.header.kid);
+  const validToken: any = jwt.verify(token, key.getPublicKey());
+  if (validToken) {
+    req.decoded = {
+      userMSId: validToken.oid,
```
Comment (Zummek) [inline server/middlewares/authenticationMiddleware.ts:16]: We should rename this to `userId` because we store MSId as normal id in db

File: server/controllers/authController.ts [hunk 2]
```diff
@@ -66,13 +66,18 @@ export const getToken = catchAsync(
   async (req: Request, res: Response) => {
     const tokenRequest = {
       code: req.body.code as string,
-      scopes: ['user.read'],
+      scopes: ['openid'],
+      nonce: 123,
```
Comment (Zummek) [inline server/controllers/authController.ts:70]: Out of curiosity, what is `nonce` and why the value of that is `123`?

File: server/middlewares/authenticationMiddleware.ts [hunk 3]
```diff
@@ -1,63 +1,24 @@
 import { Request, Response, NextFunction } from 'express';
 import * as jwt from 'jsonwebtoken';
-// import { verifyAzureToken } from 'azure-ad-jwt-lite';
```
Comment (Zummek) [inline server/middlewares/authenticationMiddleware.ts:3]: Remove this lib from package

**Review by Zummek — APPROVED** [review]
Nice work, +5 style points for desc 
look to comments, anyway lgtm

Think about updating the readme when you remember the configuration

File: server/controllers/authController.ts (hunk 2 above)
Comment (eastemployeer) [reply server/controllers/authController.ts:70]: Nonce is a field that originally is generated by MS to identify token as token only for MS Graph (so without possibility to validate). I have to add this field manually with some value to overwrite auto-generated nonce field to indicate that this token is not only for MS Graph

File: server/middlewares/authenticationMiddleware.ts (hunk 1 above)
Comment (eastemployeer) [reply server/middlewares/authenticationMiddleware.ts:16]: I think `userMSId` is reffered in several places. I would have to change this also there. I will check this today.
